File: src/state.ts
import { randomBytes } from "node:crypto";
import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { AGENTS, type AgentName, type Side, SIDES } from "./constants.js";

export interface SideMeta {
  branch: string;
  worktree: string;
}

export interface ChallengerMeta extends SideMeta {
  pid: number | null;
  started: number | null;
  finished: number | null;
  exitCode: number | null;
  log: string;
  exitFile: string;
}

export interface SessionMeta {
  session_id: string;
  version: number;
  created: number;
  host_agent: AgentName;
  challenger_agent: AgentName;
  repo_root: string;
  base_branch: string;
  base_commit: string;
  phase: string;
  host: SideMeta & { submitted: number | null };
  challenger: ChallengerMeta;
  reviews: Record<string, unknown>;
}

export const stateRoot = process.env.CODEXCODE_STATE_ROOT ?? join(homedir(), ".codexcode");
export const sessionsDir = join(stateRoot, "sessions");

export function newSessionId(): string {
  return randomBytes(5).toString("hex");
}

export function sessionPath(sessionId: string): string {
  return join(sessionsDir, sessionId);
}

export function metaPath(sessionId: string): string {
  return join(sessionPath(sessionId), "meta.json");
}

export function initSession(input: {
  sessionId: string;
  hostAgent: AgentName;
  challengerAgent: AgentName;
  prompt: string;
  repoRoot: string;
  baseBranch: string;
  baseCommit: string;
}): SessionMeta {
  if (!AGENTS.includes(input.hostAgent) || !AGENTS.includes(input.challengerAgent)) {
    throw new Error(`unknown agent name in ${input.hostAgent}, ${input.challengerAgent}`);
  }
  if (input.hostAgent === input.challengerAgent) {
    throw new Error("host and challenger must differ");
  }
  const root = sessionPath(input.sessionId);
  mkdirSync(join(root, "work"), { recursive: true });
  mkdirSync(join(root, "logs"), { recursive: true });
  mkdirSync(join(root, "reviews"), { recursive: true });
  writeFileSync(join(root, "prompt.txt"), input.prompt, "utf8");

  const meta: SessionMeta = {
    session_id: input.sessionId,
    version: 2,
    created: Date.now() / 1000,
    host_agent: input.hostAgent,
    challenger_agent: input.challengerAgent,
    repo_root: input.repoRoot,
    base_branch: input.baseBranch,
    base_commit: input.baseCommit,
    phase: "started",
    challenger: {
      pid: null,
      branch: `codexcode-${input.sessionId}-${input.challengerAgent}`,
      worktree: join(root, "work", "challenger"),
      started: null,
      finished: null,
      exitCode: null,
      log: join(root, "logs", "challenger.log"),
      exitFile: join(root, "logs", "challenger.exit"),
    },
    host: {
      branch: `codexcode-${input.sessionId}-${input.hostAgent}`,
      worktree: join(root, "work", "host"),
      submitted: null,
    },
    reviews: {
      host_of_challenger: null,
      challenger_of_host: null,
    },
  };
  saveMeta(input.sessionId, meta);
  return meta;
}

export function saveMeta(sessionId: string, meta: SessionMeta): void {
  const path = metaPath(sessionId);
  mkdirSync(sessionPath(sessionId), { recursive: true });
  const tmp = `${path}.tmp`;
  writeFileSync(tmp, `${JSON.stringify(meta, null, 2)}\n`, "utf8");
  renameSync(tmp, path);
}

export function loadMeta(sessionId: string): SessionMeta {
  const path = metaPath(sessionId);
  if (!existsSync(path)) {
    throw new Error(`session not found: ${sessionId}`);
  }
  return JSON.parse(readFileSync(path, "utf8")) as SessionMeta;
}

export function updateMeta(sessionId: string, changes: Partial<SessionMeta>): SessionMeta {
  const meta = loadMeta(sessionId);
  const next = { ...meta, ...changes };
  saveMeta(sessionId, next);
  return next;
}

export function listSessions(): string[] {
  if (!existsSync(sessionsDir)) {
    return [];
  }
  return readdirSync(sessionsDir).filter((name) => existsSync(metaPath(name)));
}

export function removeSession(sessionId: string): void {
  rmSync(sessionPath(sessionId), { recursive: true, force: true });
}

export function diffPath(sessionId: string, side: Side): string {
  if (!SIDES.includes(side)) {
    throw new Error(`unknown side: ${side}`);
  }
  return join(sessionPath(sessionId), `${side}.diff`);
}

export function reviewPath(sessionId: string, reviewerSide: Side): string {
  if (!SIDES.includes(reviewerSide)) {
    throw new Error(`unknown reviewer side: ${reviewerSide}`);
  }
  return join(sessionPath(sessionId), "reviews", `by_${reviewerSide}.md`);
}

export function promptPath(sessionId: string): string {
  return join(sessionPath(sessionId), "prompt.txt");
}

export function artifactPath(sessionId: string): string {
  return join(sessionPath(sessionId), "artifact.md");
}

export function writeText(path: string, content: string): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content, "utf8");
}

export function readText(path: string): string | null {
  if (!existsSync(path)) {
    return null;
  }
  return readFileSync(path, "utf8");
}

File: src/codexcode/state.py
"""Session state on disk.

Each invocation produces a session directory under ~/.codexcode/sessions/<id>/
containing the prompt, both worktrees, captured diffs, reviews, and metadata.

Sessions are removed when work is accepted or abandoned. They are not designed
to survive shell reboots cleanly; agents listing them after a crash should treat
them as recoverable scratch space.
"""
from __future__ import annotations

import json
import os
import pathlib
import secrets
import shutil
import time
from typing import Any, Dict, Optional

STATE_ROOT = pathlib.Path(os.path.expanduser("~/.codexcode"))
SESSIONS_DIR = STATE_ROOT / "sessions"

HOST = "host"
CHALLENGER = "challenger"
SIDES = (HOST, CHALLENGER)

AGENT_CLAUDE = "claude-code"
AGENT_CODEX = "codex"
AGENTS = (AGENT_CLAUDE, AGENT_CODEX)


def new_session_id() -> str:
    return secrets.token_hex(5)


def session_path(session_id: str) -> pathlib.Path:
    return SESSIONS_DIR / session_id


def meta_path(session_id: str) -> pathlib.Path:
    return session_path(session_id) / "meta.json"


def init_session(
    session_id: str,
    host_agent: str,
    challenger_agent: str,
    prompt: str,
    repo_root: str,
    base_branch: str,
    base_commit: str,
) -> Dict[str, Any]:
    if host_agent not in AGENTS or challenger_agent not in AGENTS:
        raise ValueError(f"unknown agent name in {host_agent!r}, {challenger_agent!r}")
    if host_agent == challenger_agent:
        raise ValueError("host and challenger must differ")
    sp = session_path(session_id)
    sp.mkdir(parents=True, exist_ok=False)
    (sp / "work").mkdir()
    (sp / "logs").mkdir()
    (sp / "reviews").mkdir()
    (sp / "prompt.txt").write_text(prompt, encoding="utf-8")
    meta: Dict[str, Any] = {
        "session_id": session_id,
        "version": 1,
        "created": time.time(),
        "host_agent": host_agent,
        "challenger_agent": challenger_agent,
        "repo_root": repo_root,
        "base_branch": base_branch,
        "base_commit": base_commit,
        "phase": "started",
        "challenger": {
            "pid": None,
            "branch": f"codexcode-{session_id}-{challenger_agent}",
            "worktree": str(sp / "work" / "challenger"),
            "started": None,
            "finished": None,
            "exit_code": None,
            "log": str(sp / "logs" / "challenger.log"),
        },
        "host": {
            "branch": f"codexcode-{session_id}-{host_agent}",
            "worktree": str(sp / "work" / "host"),
            "submitted": None,
        },
        "reviews": {
            "host_of_challenger": None,
            "challenger_of_host": None,
        },
    }
    save_meta(session_id, meta)
    return meta


def save_meta(session_id: str, meta: Dict[str, Any]) -> None:
    path = meta_path(session_id)
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    tmp.write_text(json.dumps(meta, indent=2), encoding="utf-8")
    tmp.replace(path)


def load_meta(session_id: str) -> Dict[str, Any]:
    path = meta_path(session_id)
    if not path.exists():
        raise FileNotFoundError(f"session not found: {session_id}")
    return json.loads(path.read_text(encoding="utf-8"))


def update_meta(session_id: str, **changes: Any) -> Dict[str, Any]:
    meta = load_meta(session_id)
    for k, v in changes.items():
        meta[k] = v
    save_meta(session_id, meta)
    return meta


def list_sessions() -> list:
    if not SESSIONS_DIR.exists():
        return []
    out = []
    for child in sorted(SESSIONS_DIR.iterdir()):
        if (child / "meta.json").exists():
            out.append(child.name)
    return out


def remove_session(session_id: str) -> None:
    sp = session_path(session_id)
    if sp.exists():
        shutil.rmtree(sp, ignore_errors=True)


def diff_path(session_id: str, side: str) -> pathlib.Path:
    if side not in SIDES:
        raise ValueError(f"unknown side: {side}")
    return session_path(session_id) / f"{side}.diff"


def review_path(session_id: str, reviewer_side: str) -> pathlib.Path:
    if reviewer_side not in SIDES:
        raise ValueError(f"unknown reviewer side: {reviewer_side}")
    return session_path(session_id) / "reviews" / f"by_{reviewer_side}.md"


def prompt_path(session_id: str) -> pathlib.Path:
    return session_path(session_id) / "prompt.txt"


def write_text(path: pathlib.Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def read_text(path: pathlib.Path) -> Optional[str]:
    if not path.exists():
        return None
    return path.read_text(encoding="utf-8")
